Improve sm service manager memory consumption

mlohvynenko opened the pull request.

mlohvynenko committed:
[common,tools] Make allocator asserts more precise

Signed-off-by: Mykhailo Lohvynenko <Mykhailo_Lohvynenko@epam.com>

mlohvynenko committed:
[common,spaceallocator] Create pointer with make unique

Use make unique to create pointer to make sure object is properly
destroyed and allocator is properly cleaned up.

Signed-off-by: Mykhailo Lohvynenko <Mykhailo_Lohvynenko@epam.com>

mlohvynenko committed:
[sm,launcher] Create pointer with make shared

Use make shared to create pointer to make sure object is properly
destroyed and allocator is properly cleaned up.

Signed-off-by: Mykhailo Lohvynenko <Mykhailo_Lohvynenko@epam.com>

mlohvynenko committed:
[sm,servicemanager] Sync get service calls with a mutex

Get service allocates service data array on the allocator, so
concurrent calls to get service can lead to allocator memory
over consumption. This patch adds a mutex to reduce
max allocator memory consumption.

Signed-off-by: Mykhailo Lohvynenko <Mykhailo_Lohvynenko@epam.com>

al1img commented on include/aos/common/spaceallocator/spaceallocator.hpp:
```diff
@@ -179,7 +179,7 @@ class SpaceAllocator : public SpaceAllocatorItf {
      */
     RetWithError<UniquePtr<SpaceItf>> AllocateSpace(uint64_t size) override
     {
-        return UniquePtr<SpaceItf>(new (&mAllocator) Space(size));
```
It should be valid case as well. We have to fix UniquePtr.

mlohvynenko replied:
Updated UniquePtr constructors.
But I kept change to use MakeUnique.

al1img commented on tests/common/src/spaceallocator/spaceallocator_test.cpp:
```diff
@@ -0,0 +1,42 @@
+/*
```
Adding spaceallocator test should be in separate commit

mlohvynenko committed:
[common,tools] Update unique ptr constructors

Currently, unique ptr constructs default deleter
object if deleter was not passed, that leads to
object not being deallocated properly.
This patch updates the unique ptr constructors
so that user should provide allocator or a deleter
as a param to the constructor.

Signed-off-by: Mykhailo Lohvynenko <Mykhailo_Lohvynenko@epam.com>
